### ISSUE TITLE
Deleting, upgrading, or otherwise updating a road makes vehicles despawn instead of recalculating their route

### ISSUE THREAD

**Issue by 22Pilots** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#218]
This is the same issue as issue #86 which was thought solved with Labs version 1.10.16. When you make any kind of adjustment to a roadway segment that would force a vehicle to recalculate its route, the vehicle will instead despawn upon reaching the adjusted segment. To make matters stranger, the issue only seems to occur in certain areas or scenarios for me. In my case, making changes to any road in a transit / parking lot complex for an amusement park triggers this bug.

![ezgif-1-2da6e31c5b3f](https://user-images.githubusercontent.com/48328204/53973771-59c4fa00-40cf-11e9-8bc5-b9ac37ef915e.gif)

Traffic eventually normalizes after all the vehicles that existed at the time of the change get filtered out or you use the 'Clear Traffic' button.

**Comment by aubergine10** [comment]
Was this outside the central 25 tiles are of the map?

**Comment by 22Pilots** [comment]
No, this occurs within the central 25 tiles for me.

I should add that the repro in the gif was done with only Network Extensions, TM:PE Labs version 1.10.16, and 81 tiles enabled.

**Comment by aubergine10** [comment]
Savegame...

[Wilmington2.zip](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/2942413/Wilmington2.zip)

@22Pilots  Could you also [send us your log file](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/wiki/Share-your-Cities-Skylines-log-file) so we can see if there's any errors, etc?

**Comment by 22Pilots** [comment]
Here you go:
[output_log.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/2942436/output_log.txt)

**Comment by aubergine10** [comment]
Mods:

```
Bundled Mod: HardMode.dll 
Bundled Mod: UnlimitedMoney.dll 
Bundled Mod: UnlimitedOilAndOre.dll 
Bundled Mod: UnlimitedSoil.dll 
Bundled Mod: UnlockAll.dll 
Workshop Mod: Roads United NA PLUS.dll 
Workshop Mod: RoadsUnited_Core.dll 
Workshop Mod: NodeElevationControl.dll 
Workshop Mod: NoPillars.dll 
Workshop Mod: FullRefundFromRoadsMod.dll 
Workshop Mod: AutoEmptyingExtended.dll 
Workshop Mod: ObjUnity3D.dll 
Workshop Mod: Transit.Framework.dll 
Workshop Mod: NetworkExtensions2.dll 
Workshop Mod: Transit.Framework.ExtensionPoints.dll 
Workshop Mod: Transit.Addon.RoadExtensions.dll 
Workshop Mod: Transit.Addon.Tools.dll 
Workshop Mod: Transit.Framework.Redirection.dll 
Workshop Mod: Transit.Framework.Hooks.dll 
Workshop Mod: SpawnPointsFix.dll 
Workshop Mod: FineRoadAnarchy.dll 
Workshop Mod: FineRoadTool.dll 
Workshop Mod: NaturalResourcesBrush.dll 
Workshop Mod: SurfacePainter.dll 
Workshop Mod: American RoadSigns.dll 
Workshop Mod: MoreBeautification.dll 
Workshop Mod: ParallelRoadTool.dll 
Workshop Mod: NewAmericanTrafficLights.dll 
Workshop Mod: RoadsUnited_US.dll 
Workshop Mod: EnvironmentChanger.dll 
Workshop Mod: WG_CitizenEdit.dll 
Workshop Mod: EightyOne.dll 
Workshop Mod: UnlimitedOutsideConnections.dll 
Workshop Mod: PrefabHook.dll 
Workshop Mod: AnyRoadOutsideConnections.dll 
Workshop Mod: 0Harmony.dll 
Workshop Mod: CSL_RebalancedIndustriesDistricts.dll 
Workshop Mod: CSLShowMoreLimits.dll 
Workshop Mod: V10Speed.dll 
Workshop Mod: AirportRoads.dll 
Workshop Mod: MoveIt.dll 
Workshop Mod: CSUtil.CameraControl.dll 
Workshop Mod: CSUtil.Commons.dll 
Workshop Mod: TMPE.GenericGameBridge.dll 
Workshop Mod: TMPE.CitiesGameBridge.dll 
Workshop Mod: TrafficManager.dll 
Workshop Mod: TreeFireControl.dll 
```

**Comment by krzychu124** [comment]
The first thing I've seen yesterday after loading that save was that vehicles gets confused while entering the parking using first entrance.
After deleting that road they despawn completely, similar like on this gif. 
Additionally I found some floating detached segments of train tracks and airplanes

**Comment by 22Pilots** [comment]
Disregard that last output log file, I had a brain fart and didn't bother to grab it from the repro environment. Here's the fixed version:
[output_log.txt](https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/files/2942493/output_log.txt)

**Comment by aubergine10** [comment]
@22Pilots : While I don't think the following is causing the issue, some of your mods are obsolete or replaced by better alternatives.

You are using outdated Roads United Core: https://steamcommunity.com/sharedfiles/filedetails/?id=726005715

> Suggestion: Upgrade to a newer version (I've never used Roads United stuff so I don't know what latest version is)

Network Nodes Experimental is broken and causes phantom nodes and other strangeness. https://steamcommunity.com/sharedfiles/filedetails/?id=658653260

> Suggestion: Use 'Move It!' instead: https://steamcommunity.com/sharedfiles/filedetails/?id=1619685021

No pillars can also cause some strangeness: https://steamcommunity.com/sharedfiles/filedetails/?id=463845891

> Suggestion: Use 'Network Skins' instead: https://steamcommunity.com/sharedfiles/filedetails/?id=543722850

Automatic Emptying Extended is obsolete: https://steamcommunity.com/sharedfiles/filedetails/?id=686588890

> Suggestion: Use 'Empty It!' instead: https://steamcommunity.com/sharedfiles/filedetails/?id=1661072176

More Beautification causes lag when opening build menus: https://steamcommunity.com/sharedfiles/filedetails/?id=505480567

> Suggestion: Use 'Find It!' instead: https://steamcommunity.com/sharedfiles/filedetails/?id=837734529

CSL Show More Limits is somewhat old mod: https://steamcommunity.com/sharedfiles/filedetails/?id=531738447

> Suggestion: Use 'Watch It!' instead which has a hugely improved limits panel: https://steamcommunity.com/sharedfiles/filedetails/?id=1643902284

Also, completely unsubscribe any mods you don't use as sometimes they still change game code even when not enabled.

**Comment by krzychu124** [comment]
I think I found possible cause of issue. I haven't checked how to fix that yet but for now set **Vehicle restrictions aggression** to _Low_ to be able play normally

**Comment by 22Pilots** [comment]
Awesome, I'll be sure to test this later. I had no idea just how many of those mods had updated replacements!

**Comment by krzychu124** [comment]
Unfortunately setting that option to low will cause small issues like this... (they'll sometimes ignore car ban on bus lanes...)

**Look at black camper van** 😉 
- Not allowed turn left?? No problem...

![zrzut ekranu 130](https://user-images.githubusercontent.com/19638970/53991669-6414e180-412b-11e9-9873-6dcd75d90e35.png)

- Turn right and and do U-turn to go straight 😄 

![zrzut ekranu 129](https://user-images.githubusercontent.com/19638970/53991668-637c4b00-412b-11e9-95c1-aaa665cb8708.png)

**Comment by 22Pilots** [comment]
Wild 😂 Nothing a little lane assignment can't fix!

**Comment by VictorPhilipp** [comment]
Confirmed: I added a lane connection (btw you have too many xD) on the 4-lane one-way and vehicles start to disappear.

**Comment by VictorPhilipp** [comment]
Have you disabled and re-enabled the Parking AI recently?

Almost all cims/cars want to  park at those large parking lots in order to get to your zoo (probably by taking the bus). The problem is, I see that they do not know anymore **why** they want to park there (Parking AI keeps track of the "why").

I cleared all traffic and let the simulation run until cars appeared again on the 4-lane one-way: When I now force route recalculation now (e.g. by adding a lane connection) cars do not despawn anymore.

If you disabled the Parking AI, the "why" information is lost - that means cars still drive towards the parking lot (a custom via point) but then forget what their original target is. If route recalculation is enforced while they are directed towards a via point (such as a parking lot) and Parking AI detects an invalid state (e.g. it does not know where to send them next), things become problematic. While Parking AI should be able to recover from such situations this might explain the despawn.

I'll have to investigate further...

**Comment by VictorPhilipp** [comment]
> 
> 
> I think I found possible cause of issue. I haven't checked how to fix that yet but for now set **Vehicle restrictions aggression** to _Low_ to be able play normally

Lowering vehicle restrictions aggression works because the road to the zoo is restricted for passenger cars:
![grafik](https://user-images.githubusercontent.com/4333935/54078938-3e760c80-42d2-11e9-8534-756fcacc9b74.png)

This restriction is the reason why Parking AI fails to recover from the lost "why" information. Internally, first a regular path-find to the target (zoo) is executed (as if the Parking AI was not active). During this path-find, vehicle restrictions are taken into account.

**Comment by krzychu124** [comment]
Oh... Good to know 😃

**Comment by 22Pilots** [comment]
I had disabled and reenabled the parking AI in response to vehicles disappearing, not prior to vehicles beginning to disappear.

**Comment by VictorPhilipp** [comment]
> 
> 
> I had disabled and reenabled the parking AI in response to vehicles disappearing, not prior to vehicles beginning to disappear.

Yes, even after making sure that the Parking AI has every information it needs they still despawn after a route recalculation.

The main problem is that the Parking AI was never designed for situations where the target is not reachable by car at all.

**Comment by 22Pilots** [comment]
Ah I see, that makes more sense. I left that connection there to act as an access road for maintenance and emergency vehicles. Would deleting the access road force the parking AI to recognize the transit option as it's only route or would it still trigger issues?

**Comment by VictorPhilipp** [comment]
If you delete the access road you will get problems with city services (police, garbage, etc.). There is no free lunch I guess :-)

I will investigate why path calculation fails and try to integrate a solution.

**Comment by 22Pilots** [comment]
Providing the park with it's own isolated services would in theory get around that I'd assume, albeit not particularly realistically but I suppose it beats the current issue!

**Comment by VictorPhilipp** [comment]
This won't work. You will get service problems anyway because the game will try to assign city-wide service tasks to your isolated service buildings which they cannot fulfill. For example, you will see things like service vehicles counters hopping from 0/20 to 1/20 and back to 0/20 and residential zones complaining about missing services.

**Comment by 22Pilots** [comment]
I see, so much for that. Looking forward to whatever fixes you can come up with then!

**Comment by VictorPhilipp** [comment]
Technical stuff ahead:

I have traced a full path-find execution which failed:
```
UNIT(215775)
PathFindImplementation: Preparing calculation:
	bufferItemStartA: segment=19233 lane=1 off=229 laneId=228806
	bufferItemStartB: segment=0 lane=0 off=0 laneId=0
	bufferItemEndA: segment=27583 lane=0 off=139 laneId=226216
	bufferItemEndB: segment=0 lane=0 off=0 laneId=0
	vehicleItem: segment=0 lane=0 off=0 laneId=0 vehiclePosIndicator=1
Properties:
	[...]
	pathType=DrivingOnly
	[...]
```

- In the current implementation the Parking AI tries to calculate a regular car path to the target building first (`pathType=DrivingOnly`).
- The start segment (19233) is located at the 4-lane one-way road
![grafik](https://user-images.githubusercontent.com/4333935/54079405-3cb14680-42dc-11e9-954e-8648380e204e.png)
- The end segment (27583) is located in front of the "House of Horros" (building 27573)
![grafik](https://user-images.githubusercontent.com/4333935/54079408-5b174200-42dc-11e9-95a0-2ce9b73fad6a.png)
- The target is only reachable by a `drive car -> walk -> monorail -> walk` path
- Using public transport is prohibited on the path (`m_laneTypes=Vehicle, Pedestrian, TransportVehicle` and `m_vehicleTypes=Car`). The monorail segment 25711 is inaccessible.

**Comment by 22Pilots** [comment]
Interesting, I think I clipped the monorail track over the corner of the House of Horrors building, I wonder if the game somehow considered it blocked because of that?

**Comment by VictorPhilipp** [comment]
> Interesting, I think I clipped the monorail track over the corner of the House of Horrors building, I wonder if the game somehow considered it blocked because of that?
No, that's not a problem.

Referencing #196, since a fix for #196 should give us a solution for this problem

**Comment by VictorPhilipp** [comment]
Also referencing #47
